lib/pure/os.nim

```diff
@@ -3187,38 +3187,59 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
         raiseOSError(osLastError(), path)
     rawToFormalFileInfo(rawInfo, path, result)
 
-proc sameFileContent*(path1, path2: string): bool {.rtl, extern: "nos$1",
-  tags: [ReadIOEffect], noWeirdTarget.} =
-  ## Returns true if both pathname arguments refer to files with identical
-  ## binary content.
+proc sameFileContent*(path1, path2: string; checkFiles = false, checkSize = false, bufferSize = 0.Natural): bool
+  {.rtl, extern: "nos$1", tags: [ReadIOEffect], noWeirdTarget.} =
+  ## Returns `true` if both pathname arguments refer to files with identical binary content.
+  ##
+  ## If `checkSize` is `true` then checks the file sizes *before reading the files*,
+  ## if the files have different file sizes they can not have the same contents,
+  ## `checkSize = true` may be faster, specially for very big files.
+  ## This does not fail if the file never existed in the first place, unless `checkFiles = true`.
   ##
   ## See also:
   ## * `sameFile proc <#sameFile,string,string>`_
-  var
-    a, b: File
-  if not open(a, path1): return false
-  if not open(b, path2):
+  runnableExamples:
+    doAssert sameFileContent(currentSourcePath, currentSourcePath, checkSize = true, bufferSize = 4096)
+  var a, b: File
+  var bufA, bufB: pointer
+  try:  # readBuffer or open may or may not raise IOError.
+    if not open(a, path1):
+      if checkFiles:
+        raise newException(IOError, "Can not open file: $1" % [path1])
+      return false
+    if not open(b, path2):
+      if checkFiles:
+        raise newException(IOError, "Can not open file: $1" % [path2])
+      return false
+
+    var bufferSize = bufferSize
+    if bufferSize == 0 or checkSize:
+      var infoA = getFileInfo(a)
+      if bufferSize == 0: bufferSize = infoA.blockSize
```

Comment:
I think this is still wrong, because `path1` and `path2` can be on different disks,
with different file system format and different block sizes ?,
maybe something like `min(infoA.blockSize, infoB.blockSize)`.
:thinking: